Can not build with legacy toolchain and type traits are not the same when enable/disable C++11's STL.

**Describe your environment**

OS: Linux
Compiler: gcc 4.8

**What is the actual behavior?**

Problem 1:

```bash
[ 52%] Building CXX object ext/src/http/client/curl/CMakeFiles/opentelemetry_http_client_curl.dir/http_client_curl.cc.o
cd /home/runner/work/cmake-toolset/cmake-toolset/test/build_jobs_dir/_deps/opentelemetry_cpp-v1.19.0/linux-x86_64-gnu-4/ext/src/http/client/curl && /usr/bin/g++-4.8 -DCURL_STATICLIB -DENABLE_ASYNC_EXPORT -DENABLE_METRICS_EXEMPLAR_PREVIEW -DENABLE_OTLP_COMPRESSION_PREVIEW -DENABLE_OTLP_GRPC_SSL_MTLS_PREVIEW -DHAVE_ABSEIL -DOPENTELEMETRY_ABI_VERSION_NO=1 -D__CLANG_SUPPORT_DYN_ANNOTATION__ -I/home/runner/work/cmake-toolset/cmake-toolset/test/build_jobs_dir/_deps/opentelemetry_cpp-v1.19.0/linux-x86_64-gnu-4/generated/third_party/opentelemetry-proto -I/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/api/include -I/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/sdk/include -I/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/sdk -I/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/ext/include -isystem /home/runner/work/cmake-toolset/cmake-toolset/test/third_party/install/linux-x86_64-gnu-4/include -fexceptions -frtti -pthread -O2 -g -DNDEBUG -ggdb -std=gnu++1y -fPIC -Wno-error=deprecated-declarations -MD -MT ext/src/http/client/curl/CMakeFiles/opentelemetry_http_client_curl.dir/http_client_curl.cc.o -MF CMakeFiles/opentelemetry_http_client_curl.dir/http_client_curl.cc.o.d -o CMakeFiles/opentelemetry_http_client_curl.dir/http_client_curl.cc.o -c /home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/ext/src/http/client/curl/http_client_curl.cc
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/ext/src/http/client/curl/http_client_curl.cc: In function ‘int opentelemetry::v1::ext::http::client::curl::deflateInPlace(z_stream*, unsigned char*, uint32_t, uint32_t*)’:
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/ext/src/http/client/curl/http_client_curl.cc:119:27: error: expected type-specifier
   auto hold = static_cast<std::remove_const_t<decltype(z_stream::next_in)>>(
                           ^
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/ext/src/http/client/curl/http_client_curl.cc:119:27: error: expected ‘>’
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/ext/src/http/client/curl/http_client_curl.cc:119:27: error: expected ‘(’
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/ext/src/http/client/curl/http_client_curl.cc:119:27: error: ‘remove_const_t’ is not a member of ‘std’
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/ext/src/http/client/curl/http_client_curl.cc:119:47: error: expected primary-expression before ‘decltype’
   auto hold = static_cast<std::remove_const_t<decltype(z_stream::next_in)>>(
```

Problem 2:

```bash
[ 96%] Building CXX object exporters/elasticsearch/CMakeFiles/opentelemetry_exporter_elasticsearch_logs.dir/src/es_log_recordable.cc.o
cd /home/runner/work/cmake-toolset/cmake-toolset/test/build_jobs_dir/_deps/opentelemetry_cpp-v1.19.0/linux-x86_64-gnu-4/exporters/elasticsearch && /usr/bin/g++-4.8 -DENABLE_ASYNC_EXPORT -DENABLE_METRICS_EXEMPLAR_PREVIEW -DENABLE_OTLP_COMPRESSION_PREVIEW -DENABLE_OTLP_GRPC_SSL_MTLS_PREVIEW -DHAVE_ABSEIL -DOPENTELEMETRY_ABI_VERSION_NO=1 -D__CLANG_SUPPORT_DYN_ANNOTATION__ -I/home/runner/work/cmake-toolset/cmake-toolset/test/build_jobs_dir/_deps/opentelemetry_cpp-v1.19.0/linux-x86_64-gnu-4/generated/third_party/opentelemetry-proto -I/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/api/include -I/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/sdk/include -I/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/sdk -I/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/ext/include -I/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/exporters/elasticsearch/include -isystem /home/runner/work/cmake-toolset/cmake-toolset/test/third_party/install/linux-x86_64-gnu-4/include -fexceptions -frtti -pthread -O2 -g -DNDEBUG -ggdb -std=gnu++1y -fPIC -Wno-error=deprecated-declarations -MD -MT exporters/elasticsearch/CMakeFiles/opentelemetry_exporter_elasticsearch_logs.dir/src/es_log_recordable.cc.o -MF CMakeFiles/opentelemetry_exporter_elasticsearch_logs.dir/src/es_log_recordable.cc.o.d -o CMakeFiles/opentelemetry_exporter_elasticsearch_logs.dir/src/es_log_recordable.cc.o -c /home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/exporters/elasticsearch/src/es_log_recordable.cc
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/exporters/elasticsearch/src/es_log_recordable.cc: In static member function ‘static void nlohmann::json_abi_v3_11_3::adl_serializer<absl::lts_2020_02_25::variant<bool, int, unsigned int, long int, double, std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<bool, std::allocator<bool> >, std::vector<int, std::allocator<int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long int, std::allocator<long int> >, std::vector<double, std::allocator<double> >, std::vector<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, long unsigned int, std::vector<long unsigned int, std::allocator<long unsigned int> >, std::vector<unsigned char, std::allocator<unsigned char> > > >::to_json(nlohmann::json_abi_v3_11_3::json&, const OwnedAttributeValue&)’:
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/exporters/elasticsearch/src/es_log_recordable.cc:31:50: error: parameter declared ‘auto’
     opentelemetry::nostd::visit([&j](const auto &value) { j = value; }, v);
                                                  ^
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/exporters/elasticsearch/src/es_log_recordable.cc: In lambda function:
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/exporters/elasticsearch/src/es_log_recordable.cc:31:63: error: ‘value’ was not declared in this scope
     opentelemetry::nostd::visit([&j](const auto &value) { j = value; }, v);
                                                               ^
In file included from /home/runner/work/cmake-toolset/cmake-toolset/test/third_party/install/linux-x86_64-gnu-4/include/absl/base/internal/invoke.h:42:0,
                 from /home/runner/work/cmake-toolset/cmake-toolset/test/third_party/install/linux-x86_64-gnu-4/include/absl/utility/utility.h:50,
                 from /home/runner/work/cmake-toolset/cmake-toolset/test/third_party/install/linux-x86_64-gnu-4/include/absl/types/variant.h:46,
                 from /home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/api/include/opentelemetry/nostd/variant.h:55,
                 from /home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/api/include/opentelemetry/common/attribute_value.h:10,
                 from /home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/sdk/include/opentelemetry/sdk/common/attribute_utils.h:14,
                 from /home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/exporters/elasticsearch/include/opentelemetry/exporters/elasticsearch/es_log_recordable.h:13,
                 from /home/runner/work/cmake-toolset/cmake-toolset/test/third_party/packages/opentelemetry-cpp-v1.19.0/exporters/elasticsearch/src/es_log_recordable.cc:15:
/home/runner/work/cmake-toolset/cmake-toolset/test/third_party/install/linux-x86_64-gnu-4/include/absl/meta/type_traits.h: In substitution of ‘template<class T> using result_of_t = typename std::result_of::type [with T = nlohmann::json_abi_v3_11_3::adl_serializer<absl::lts_2020_02_25::variant<bool, int, unsigned int, long int, double, std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<bool, std::allocator<bool> >, std::vector<int, std::allocator<int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long int, std::allocator<long int> >, std::vector<double, std::allocator<double> >, std::vector<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, long unsigned int, std::vector<long unsigned int, std::allocator<long unsigned int> >, std::vector<unsigned char, std::allocator<unsigned char> > > >::to_json(nlohmann::json_abi_v3_11_3::json&, const OwnedAttributeValue&)::__lambda29(absl::lts_2020_02_25::variant_internal::VariantAccessResult<0ul, const absl::lts_2020_02_25::variant<bool, int, unsigned int, long int, double, std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<bool, std::allocator<bool> >, std::vector<int, std::allocator<int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long int, std::allocator<long int> >, std::vector<double, std::allocator<double> >, std::vector<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, long unsigned int, std::vector<long unsigned int, std::allocator<long unsigned int> >, std::vector<unsigned char, std::allocator<unsigned char> > >&>)]’:
```

--- a/api/include/opentelemetry/nostd/type_traits.h
+++ b/api/include/opentelemetry/nostd/type_traits.h
@@ -17,6 +17,7 @@
 #  include "opentelemetry/config.h"
 #  include "opentelemetry/nostd/detail/void.h"  // IWYU pragma: export
 #  include "opentelemetry/version.h"
+#endif
 
 OPENTELEMETRY_BEGIN_NAMESPACE
 namespace nostd
@@ -70,6 +71,10 @@ struct remove_all_extents<std::array<T, N>> : remove_all_extents<T>
 template <typename T>
 using remove_all_extents_t = typename remove_all_extents<T>::type;
 
+#if defined(OPENTELEMETRY_STL_VERSION) && OPENTELEMETRY_STL_VERSION >= 2017
+using std::is_nothrow_swappable;
+using std::is_swappable;
+#else
 /**
  * Back port of std::is_swappable
  */
@@ -119,6 +124,7 @@ struct is_nothrow_swappable<false, T> : std::false_type
 }  // namespace detail
 template <typename T>
 using is_nothrow_swappable = detail::swappable::is_nothrow_swappable<is_swappable<T>::value, T>;
+#endif
 
 /**
  * Back port of
@@ -127,12 +133,12 @@ using is_nothrow_swappable = detail::swappable::is_nothrow_swappable<is_swappabl
  *  std::is_trivialy_copy_assignable
  *  std::is_trivialy_move_assignable
  */
-#  ifdef OPENTELEMETRY_TRIVIALITY_TYPE_TRAITS
+#ifdef OPENTELEMETRY_TRIVIALITY_TYPE_TRAITS
 using std::is_trivially_copy_assignable;
 using std::is_trivially_copy_constructible;
 using std::is_trivially_move_assignable;
 using std::is_trivially_move_constructible;
-#  else
+#else
 template <typename T>
 struct is_trivially_copy_constructible
 {
@@ -156,7 +162,6 @@ struct is_trivially_move_assignable
 {
   static constexpr bool value = __is_trivial(T);
 };
-#  endif
+#endif
 }  // namespace nostd
 OPENTELEMETRY_END_NAMESPACE
-#endif /* OPENTELEMETRY_HAVE_STD_TYPE_TRAITS */

--- a/api/include/opentelemetry/std/type_traits.h
+++ b/api/include/opentelemetry/std/type_traits.h
@@ -5,18 +5,6 @@
 
 // IWYU pragma: private, include "opentelemetry/nostd/type_traits.h"
 
-#include <type_traits>
+#include <type_traits>  // IWYU pragma: keep
 
-#include "opentelemetry/version.h"
-
-OPENTELEMETRY_BEGIN_NAMESPACE
-// Standard Type aliases in nostd namespace
-namespace nostd
-{
-
-// nostd::enable_if_t<...>
-template <bool B, class T = void>
-using enable_if_t = typename std::enable_if<B, T>::type;
-
-}  // namespace nostd
-OPENTELEMETRY_END_NAMESPACE
+#include "opentelemetry/version.h"  // IWYU pragma: keep

--- a/exporters/elasticsearch/src/es_log_recordable.cc
+++ b/exporters/elasticsearch/src/es_log_recordable.cc
@@ -23,12 +23,25 @@
 
 namespace nlohmann
 {
+template <class T>
+struct json_assign_visitor
+{
+  T *j_;
+  json_assign_visitor(T &j) : j_(&j) {}
+
+  template <class U>
+  void operator()(const U &u)
+  {
+    *j_ = u;
+  }
+};
+
 template <>
 struct adl_serializer<opentelemetry::sdk::common::OwnedAttributeValue>
 {
   static void to_json(json &j, const opentelemetry::sdk::common::OwnedAttributeValue &v)
   {
-    opentelemetry::nostd::visit([&j](const auto &value) { j = value; }, v);
+    opentelemetry::nostd::visit(json_assign_visitor<json>(j), v);
   }
 };
 
@@ -37,7 +50,7 @@ struct adl_serializer<opentelemetry::common::AttributeValue>
 {
   static void to_json(json &j, const opentelemetry::common::AttributeValue &v)
   {
-    opentelemetry::nostd::visit([&j](const auto &value) { j = value; }, v);
+    opentelemetry::nostd::visit(json_assign_visitor<json>(j), v);
   }
 };
 }  // namespace nlohmann

--- a/ext/src/http/client/curl/http_client_curl.cc
+++ b/ext/src/http/client/curl/http_client_curl.cc
@@ -15,7 +15,6 @@
 #include <mutex>
 #include <string>
 #include <thread>
-#include <type_traits>
 #include <unordered_map>
 #include <unordered_set>
 #include <utility>

--- a/ext/src/http/client/curl/http_client_curl.cc
+++ b/ext/src/http/client/curl/http_client_curl.cc
@@ -26,6 +26,7 @@
 #include "opentelemetry/ext/http/common/url_parser.h"
 #include "opentelemetry/nostd/shared_ptr.h"
 #include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/nostd/type_traits.h"
 #include "opentelemetry/sdk/common/thread_instrumentation.h"
 #include "opentelemetry/version.h"
 
@@ -116,7 +117,7 @@ int deflateInPlace(z_stream *strm, unsigned char *buf, uint32_t len, uint32_t *m
   // now empty input buffer (this will only occur for long incompressible streams, more than ~20 MB
   // for the default deflate memLevel of 8, or when *max_len is too small and less than the length
   // of the header plus one byte)
-  auto hold = static_cast<std::remove_const_t<decltype(z_stream::next_in)>>(
+  auto hold = static_cast<nostd::remove_const_t<decltype(z_stream::next_in)>>(
       strm->zalloc(strm->opaque, strm->avail_in, 1));  // allocated buffer to hold input data
   if (hold == Z_NULL)
   {

